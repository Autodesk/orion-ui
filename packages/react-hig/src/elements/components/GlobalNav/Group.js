/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import createComponent from '../../../adapters/createComponent';
import HIGElement from '../../HIGElement';
import HIGNodeList from '../../HIGNodeList';
import HIGChildValidator from '../../HIGChildValidator';
import ModuleComponent, { Module } from './Module';

export class Group extends HIGElement {
  constructor(HIGConstructor, initialProps) {
    super(HIGConstructor, initialProps);
    this.modules = new HIGNodeList({
      type: Module,
      HIGConstructor: this.hig.partials.Module,
      onAdd: (instance, beforeInstance) => {
        this.hig.addModule(instance, beforeInstance);
      }
    });

    this.state = {};
    this._render = this._render.bind(this);
  }

  componentDidMount() {
    this.modules.componentDidMount();
    this._render();
  }

  createElement(ElementConstructor, props) {
    return this.modules.createElement(ElementConstructor, props);
  }

  insertBefore(instance, insertBeforeIndex) {
    this.modules.insertBefore(instance, insertBeforeIndex);
    if (this.mounted) {
      this._render();
    }
  }

  removeChild(instance) {
    this.modules.removeChild(instance);
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    this.processUpdateProps(updatePayload).then(this._render);
  }

  isVisible() {
    return this.state.isVisible;
  }

  expandModules() {
    this.modules.forEach(module => {
      module.commitUpdate({query: undefined})
      module.expandAll();
    });
    this._render();
  }

  collapseModules() {
    this.modules.forEach(module => {
      this.commitUpdate({query: undefined})
      module.collapseAll();
    });
    this._render();
  }

  _render() {
    const matches = this.modules.map(module => {
      module.commitUpdate({
        query: this.props.query,
        activeModule: this.props.activeModule,
        activeSubmodule: this.props.activeSubmodule
      });

      return module.isVisible();
    });

    if (matches.some(m => m)) {
      this.hig.show();
      this.state.isVisible = true;
    } else {
      this.hig.hide();
      this.state.isVisible = false;
    }
  }
}

const GroupComponent = createComponent(Group);

GroupComponent.propTypes = {
  children: HIGChildValidator([ModuleComponent])
};

GroupComponent.__docgenInfo = {
  props: {
    children: {
      description: 'support adding Module'
    }
  }
};

GroupComponent.Module = ModuleComponent;

export default GroupComponent;
