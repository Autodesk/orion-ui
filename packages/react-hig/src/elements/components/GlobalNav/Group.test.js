/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import { mount } from 'enzyme';
import * as HIG from 'hig.web';
import React from 'react';

import GlobalNav from './GlobalNav';
import Section from './Section';
import SectionList from './SectionList';
import Group from './Group';
import Module from './Module';

const Context = props => {
  const { children, ...rest } = props;

  return (
    <GlobalNav>
      <GlobalNav.SideNav>
        <GlobalNav.SideNav.SectionList>
          <GlobalNav.SideNav.SectionList.Section>
            <Group>{props.children}</Group>
          </GlobalNav.SideNav.SectionList.Section>
        </GlobalNav.SideNav.SectionList>
      </GlobalNav.SideNav>
    </GlobalNav>
  );
};

function higContext() {
  const higContainer = document.createElement('div');

  // use spread here to clone defaults since HIG.Button mutates this object
  const higNav = new HIG.GlobalNav();

  higNav.mount(higContainer);

  const higSideNav = new higNav.partials.SideNav();
  higNav.addSideNav(higSideNav);

  const higSection = new higSideNav.partials.Section({});

  higSideNav.addSection(higSection);

  const collapse1 = new higSection.partials.Collapse({});

  higSection.addCollapse(collapse1);
  collapse1.minimize();

  const higGroup = new higSection.partials.Group();

  higSection.addGroup(higGroup);

  return { higNav, higSideNav, higSection, higGroup, higContainer };
}

describe('<Group>', () => {
  describe('children: <Module>', () => {
    it('can render a list of <Module> elements', () => {
      const { higGroup, higContainer } = higContext();

      const module1Defaults = {
        icon: 'project-management',
        title: 'Module 1',
        link: '#'
      };

      const module1 = new higGroup.partials.Module(module1Defaults);
      higGroup.addModule(module1);

      const module2Defaults = {
        icon: 'project-management',
        title: 'Module 2',
        link: '#'
      };
      const module2 = new higGroup.partials.Module(module2Defaults);
      higGroup.addModule(module2);

      const reactContainer = document.createElement('div');

      const wrapper = mount(
        <Context>
          <Module {...module1Defaults} />
          <Module {...module2Defaults} />
        </Context>,
        {
          attachTo: reactContainer
        }
      );

      expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();

      expect(reactContainer.firstChild.outerHTML).toEqual(
        higContainer.firstChild.outerHTML
      );
    });

    it('can insert <Module> elements before and after another <Module>', () => {
      const { higGroup, higContainer } = higContext();

      const module1Defaults = {
        icon: 'project-management',
        title: 'Module 1',
        link: '#',
        key: '1'
      };
      const module1 = new higGroup.partials.Module(module1Defaults);

      // Do not add yet

      const module2Defaults = {
        icon: 'project-management',
        title: 'Module 2',
        link: '#',
        key: '2'
      };
      const module2 = new higGroup.partials.Module(module2Defaults);
      higGroup.addModule(module2);

      // Add before item 2
      higGroup.addModule(module1, module2);

      class CustomComponent extends React.Component {
        constructor(props) {
          super(props);
          this.state = { showingItemBefore: false, showingItemAfter: false };
        }

        render() {
          return (
            <Context>
              {this.state.showingItemBefore && <Module {...module1Defaults} />}
              <Module {...module2Defaults} />
              {this.state.showingItemAfter && <Module icon="assets" />}
            </Context>
          );
        }
      }

      const reactContainer = document.createElement('div');
      const wrapper = mount(<CustomComponent />, {
        attachTo: reactContainer
      });

      expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();

      wrapper.setState({ showingItemBefore: true });

      expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();

      expect(reactContainer.firstChild.outerHTML).toEqual(
        higContainer.firstChild.outerHTML
      );

      wrapper.setState({ showingItemBefore: false });

      expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();

      wrapper.setState({ showingItemAfter: true });

      expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();

      wrapper.setState({ showingItemAfter: false });

      expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();
    });

    it('can not render HTML elements as children', () => {
      global.console.error = jest.fn();

      mount(
        <Context>
          <div>Hello world!</div>
        </Context>
      );

      expect(console.error).toBeCalledWith(
        expect.stringMatching(
          /'div' is not a valid child of Group. Children should be of type 'Module'/
        )
      );
    });

    it('can not render HTML text as children', () => {
      global.console.error = jest.fn();

      mount(
        <Context>
          Hello world!
        </Context>
      );

      expect(console.error).toBeCalledWith(
        expect.stringMatching(
          /'Hello world!' is not a valid child of Group. Children should be of type 'Module'/
        )
      );
    });
  });
});
