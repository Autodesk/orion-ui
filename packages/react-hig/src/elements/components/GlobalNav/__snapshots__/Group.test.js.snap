// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Group> children: <Item> can insert <Item> elements before and after another <Item> 1`] = `
"<div class=\\"hig__global-nav\\">
    <div class=\\"hig__global-nav__sidenav\\">
    <div class=\\"hig__global-nav__side-nav__section\\">
    <div class=\\"hig__global-nav__side-nav__section__header\\">
        <span class=\\"hig__global-nav__side-nav__section__header__label\\">Project</span>
        <span class=\\"hig__global-nav__side-nav__section__header__name\\">X</span>
    </div>
    <div class=\\"hig__global-nav__side-nav__section__group\\">
    <div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"21\\" viewBox=\\"0 0 20 21\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"project-management\\"><g transform=\\"translate(.242)\\" fill=\\"#34495E\\" fill-rule=\\"evenodd\\"><path d=\\"M15,8.15002717 C14.6738533,8.08382289 14.3400214,8.03878262 14,8.01640228 L14,1 L1,1 L1,17 L6.15002717,17 C6.21952655,17.3423796 6.31234911,17.6762897 6.42676427,18 L0,18 L0,0 L15,0 L15,8.15002717 Z M8.40093151,10 C8.07157864,10.3054873 7.76971094,10.6401947 7.49945072,11 L3,11 L3,10 L8.40093151,10 Z M6.42676427,13 C6.31234911,13.3237103 6.21952655,13.6576204 6.15002717,14 L3,14 L3,13 L6.42676427,13 Z M3,4 L12,4 L12,5 L3,5 L3,4 Z M3,7 L12,7 L12,8 L3,8 L3,7 Z\\" fill-rule=\\"nonzero\\"></path><path d=\\"M13.5,20 C15.9852814,20 18,17.9852814 18,15.5 C18,13.0147186 15.9852814,11 13.5,11 C11.0147186,11 9,13.0147186 9,15.5 C9,17.9852814 11.0147186,20 13.5,20 Z M13.5,21 C10.4624339,21 8,18.5375661 8,15.5 C8,12.4624339 10.4624339,10 13.5,10 C16.5375661,10 19,12.4624339 19,15.5 C19,18.5375661 16.5375661,21 13.5,21 Z\\" fill-rule=\\"nonzero\\"></path><rect x=\\"13\\" y=\\"12\\" width=\\"1\\" height=\\"4\\"></rect><rect x=\\"13\\" y=\\"15\\" width=\\"4\\" height=\\"1\\"></rect></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">Item 2</div>
    </a>
</div><!--ITEM-->
</div><!--GROUP-->
</div><!--SECTION-->
    <div class=\\"hig__global-nav__sidenav__links\\">
        <!--LINK-->
    </div>
    <div class=\\"hig__global-nav__sidenav__copywright\\">© 2017 Autodesk, Inc. </div>
</div><!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__container__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;

exports[`<Group> children: <Item> can insert <Item> elements before and after another <Item> 2`] = `
"<div class=\\"hig__global-nav\\">
    <div class=\\"hig__global-nav__sidenav\\">
    <div class=\\"hig__global-nav__side-nav__section\\">
    <div class=\\"hig__global-nav__side-nav__section__header\\">
        <span class=\\"hig__global-nav__side-nav__section__header__label\\">Project</span>
        <span class=\\"hig__global-nav__side-nav__section__header__name\\">X</span>
    </div>
    <div class=\\"hig__global-nav__side-nav__section__group\\">
    <div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"21\\" viewBox=\\"0 0 20 21\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"project-management\\"><g transform=\\"translate(.242)\\" fill=\\"#34495E\\" fill-rule=\\"evenodd\\"><path d=\\"M15,8.15002717 C14.6738533,8.08382289 14.3400214,8.03878262 14,8.01640228 L14,1 L1,1 L1,17 L6.15002717,17 C6.21952655,17.3423796 6.31234911,17.6762897 6.42676427,18 L0,18 L0,0 L15,0 L15,8.15002717 Z M8.40093151,10 C8.07157864,10.3054873 7.76971094,10.6401947 7.49945072,11 L3,11 L3,10 L8.40093151,10 Z M6.42676427,13 C6.31234911,13.3237103 6.21952655,13.6576204 6.15002717,14 L3,14 L3,13 L6.42676427,13 Z M3,4 L12,4 L12,5 L3,5 L3,4 Z M3,7 L12,7 L12,8 L3,8 L3,7 Z\\" fill-rule=\\"nonzero\\"></path><path d=\\"M13.5,20 C15.9852814,20 18,17.9852814 18,15.5 C18,13.0147186 15.9852814,11 13.5,11 C11.0147186,11 9,13.0147186 9,15.5 C9,17.9852814 11.0147186,20 13.5,20 Z M13.5,21 C10.4624339,21 8,18.5375661 8,15.5 C8,12.4624339 10.4624339,10 13.5,10 C16.5375661,10 19,12.4624339 19,15.5 C19,18.5375661 16.5375661,21 13.5,21 Z\\" fill-rule=\\"nonzero\\"></path><rect x=\\"13\\" y=\\"12\\" width=\\"1\\" height=\\"4\\"></rect><rect x=\\"13\\" y=\\"15\\" width=\\"4\\" height=\\"1\\"></rect></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">Item 1</div>
    </a>
</div><div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"21\\" viewBox=\\"0 0 20 21\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"project-management\\"><g transform=\\"translate(.242)\\" fill=\\"#34495E\\" fill-rule=\\"evenodd\\"><path d=\\"M15,8.15002717 C14.6738533,8.08382289 14.3400214,8.03878262 14,8.01640228 L14,1 L1,1 L1,17 L6.15002717,17 C6.21952655,17.3423796 6.31234911,17.6762897 6.42676427,18 L0,18 L0,0 L15,0 L15,8.15002717 Z M8.40093151,10 C8.07157864,10.3054873 7.76971094,10.6401947 7.49945072,11 L3,11 L3,10 L8.40093151,10 Z M6.42676427,13 C6.31234911,13.3237103 6.21952655,13.6576204 6.15002717,14 L3,14 L3,13 L6.42676427,13 Z M3,4 L12,4 L12,5 L3,5 L3,4 Z M3,7 L12,7 L12,8 L3,8 L3,7 Z\\" fill-rule=\\"nonzero\\"></path><path d=\\"M13.5,20 C15.9852814,20 18,17.9852814 18,15.5 C18,13.0147186 15.9852814,11 13.5,11 C11.0147186,11 9,13.0147186 9,15.5 C9,17.9852814 11.0147186,20 13.5,20 Z M13.5,21 C10.4624339,21 8,18.5375661 8,15.5 C8,12.4624339 10.4624339,10 13.5,10 C16.5375661,10 19,12.4624339 19,15.5 C19,18.5375661 16.5375661,21 13.5,21 Z\\" fill-rule=\\"nonzero\\"></path><rect x=\\"13\\" y=\\"12\\" width=\\"1\\" height=\\"4\\"></rect><rect x=\\"13\\" y=\\"15\\" width=\\"4\\" height=\\"1\\"></rect></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">Item 2</div>
    </a>
</div><!--ITEM-->
</div><!--GROUP-->
</div><!--SECTION-->
    <div class=\\"hig__global-nav__sidenav__links\\">
        <!--LINK-->
    </div>
    <div class=\\"hig__global-nav__sidenav__copywright\\">© 2017 Autodesk, Inc. </div>
</div><!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__container__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;

exports[`<Group> children: <Item> can insert <Item> elements before and after another <Item> 3`] = `
"<div class=\\"hig__global-nav\\">
    <div class=\\"hig__global-nav__sidenav\\">
    <div class=\\"hig__global-nav__side-nav__section\\">
    <div class=\\"hig__global-nav__side-nav__section__header\\">
        <span class=\\"hig__global-nav__side-nav__section__header__label\\">Project</span>
        <span class=\\"hig__global-nav__side-nav__section__header__name\\">X</span>
    </div>
    <div class=\\"hig__global-nav__side-nav__section__group\\">
    <div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"21\\" viewBox=\\"0 0 20 21\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"project-management\\"><g transform=\\"translate(.242)\\" fill=\\"#34495E\\" fill-rule=\\"evenodd\\"><path d=\\"M15,8.15002717 C14.6738533,8.08382289 14.3400214,8.03878262 14,8.01640228 L14,1 L1,1 L1,17 L6.15002717,17 C6.21952655,17.3423796 6.31234911,17.6762897 6.42676427,18 L0,18 L0,0 L15,0 L15,8.15002717 Z M8.40093151,10 C8.07157864,10.3054873 7.76971094,10.6401947 7.49945072,11 L3,11 L3,10 L8.40093151,10 Z M6.42676427,13 C6.31234911,13.3237103 6.21952655,13.6576204 6.15002717,14 L3,14 L3,13 L6.42676427,13 Z M3,4 L12,4 L12,5 L3,5 L3,4 Z M3,7 L12,7 L12,8 L3,8 L3,7 Z\\" fill-rule=\\"nonzero\\"></path><path d=\\"M13.5,20 C15.9852814,20 18,17.9852814 18,15.5 C18,13.0147186 15.9852814,11 13.5,11 C11.0147186,11 9,13.0147186 9,15.5 C9,17.9852814 11.0147186,20 13.5,20 Z M13.5,21 C10.4624339,21 8,18.5375661 8,15.5 C8,12.4624339 10.4624339,10 13.5,10 C16.5375661,10 19,12.4624339 19,15.5 C19,18.5375661 16.5375661,21 13.5,21 Z\\" fill-rule=\\"nonzero\\"></path><rect x=\\"13\\" y=\\"12\\" width=\\"1\\" height=\\"4\\"></rect><rect x=\\"13\\" y=\\"15\\" width=\\"4\\" height=\\"1\\"></rect></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">Item 2</div>
    </a>
</div><!--ITEM-->
</div><!--GROUP-->
</div><!--SECTION-->
    <div class=\\"hig__global-nav__sidenav__links\\">
        <!--LINK-->
    </div>
    <div class=\\"hig__global-nav__sidenav__copywright\\">© 2017 Autodesk, Inc. </div>
</div><!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__container__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;

exports[`<Group> children: <Item> can insert <Item> elements before and after another <Item> 4`] = `
"<div class=\\"hig__global-nav\\">
    <div class=\\"hig__global-nav__sidenav\\">
    <div class=\\"hig__global-nav__side-nav__section\\">
    <div class=\\"hig__global-nav__side-nav__section__header\\">
        <span class=\\"hig__global-nav__side-nav__section__header__label\\">Project</span>
        <span class=\\"hig__global-nav__side-nav__section__header__name\\">X</span>
    </div>
    <div class=\\"hig__global-nav__side-nav__section__group\\">
    <div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"21\\" viewBox=\\"0 0 20 21\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"project-management\\"><g transform=\\"translate(.242)\\" fill=\\"#34495E\\" fill-rule=\\"evenodd\\"><path d=\\"M15,8.15002717 C14.6738533,8.08382289 14.3400214,8.03878262 14,8.01640228 L14,1 L1,1 L1,17 L6.15002717,17 C6.21952655,17.3423796 6.31234911,17.6762897 6.42676427,18 L0,18 L0,0 L15,0 L15,8.15002717 Z M8.40093151,10 C8.07157864,10.3054873 7.76971094,10.6401947 7.49945072,11 L3,11 L3,10 L8.40093151,10 Z M6.42676427,13 C6.31234911,13.3237103 6.21952655,13.6576204 6.15002717,14 L3,14 L3,13 L6.42676427,13 Z M3,4 L12,4 L12,5 L3,5 L3,4 Z M3,7 L12,7 L12,8 L3,8 L3,7 Z\\" fill-rule=\\"nonzero\\"></path><path d=\\"M13.5,20 C15.9852814,20 18,17.9852814 18,15.5 C18,13.0147186 15.9852814,11 13.5,11 C11.0147186,11 9,13.0147186 9,15.5 C9,17.9852814 11.0147186,20 13.5,20 Z M13.5,21 C10.4624339,21 8,18.5375661 8,15.5 C8,12.4624339 10.4624339,10 13.5,10 C16.5375661,10 19,12.4624339 19,15.5 C19,18.5375661 16.5375661,21 13.5,21 Z\\" fill-rule=\\"nonzero\\"></path><rect x=\\"13\\" y=\\"12\\" width=\\"1\\" height=\\"4\\"></rect><rect x=\\"13\\" y=\\"15\\" width=\\"4\\" height=\\"1\\"></rect></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">Item 2</div>
    </a>
</div><div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"30\\" height=\\"30\\" viewBox=\\"0 0 30 30\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"hamburger\\"><g transform=\\"translate(-92 -185)\\" fill=\\"none\\" fill-rule=\\"evenodd\\"><rect x=\\"92\\" y=\\"185\\" width=\\"30\\" height=\\"30\\"></rect><path d=\\"M100,195 L115,195 L115,196 L100,196 L100,195 Z M100,200 L115,200 L115,201 L100,201 L100,200 Z M100,205 L115,205 L115,206 L100,206 L100,205 Z\\" fill=\\"#34495E\\"></path></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">title</div>
    </a>
</div><!--ITEM-->
</div><!--GROUP-->
</div><!--SECTION-->
    <div class=\\"hig__global-nav__sidenav__links\\">
        <!--LINK-->
    </div>
    <div class=\\"hig__global-nav__sidenav__copywright\\">© 2017 Autodesk, Inc. </div>
</div><!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__container__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;

exports[`<Group> children: <Item> can insert <Item> elements before and after another <Item> 5`] = `
"<div class=\\"hig__global-nav\\">
    <div class=\\"hig__global-nav__sidenav\\">
    <div class=\\"hig__global-nav__side-nav__section\\">
    <div class=\\"hig__global-nav__side-nav__section__header\\">
        <span class=\\"hig__global-nav__side-nav__section__header__label\\">Project</span>
        <span class=\\"hig__global-nav__side-nav__section__header__name\\">X</span>
    </div>
    <div class=\\"hig__global-nav__side-nav__section__group\\">
    <div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"21\\" viewBox=\\"0 0 20 21\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"project-management\\"><g transform=\\"translate(.242)\\" fill=\\"#34495E\\" fill-rule=\\"evenodd\\"><path d=\\"M15,8.15002717 C14.6738533,8.08382289 14.3400214,8.03878262 14,8.01640228 L14,1 L1,1 L1,17 L6.15002717,17 C6.21952655,17.3423796 6.31234911,17.6762897 6.42676427,18 L0,18 L0,0 L15,0 L15,8.15002717 Z M8.40093151,10 C8.07157864,10.3054873 7.76971094,10.6401947 7.49945072,11 L3,11 L3,10 L8.40093151,10 Z M6.42676427,13 C6.31234911,13.3237103 6.21952655,13.6576204 6.15002717,14 L3,14 L3,13 L6.42676427,13 Z M3,4 L12,4 L12,5 L3,5 L3,4 Z M3,7 L12,7 L12,8 L3,8 L3,7 Z\\" fill-rule=\\"nonzero\\"></path><path d=\\"M13.5,20 C15.9852814,20 18,17.9852814 18,15.5 C18,13.0147186 15.9852814,11 13.5,11 C11.0147186,11 9,13.0147186 9,15.5 C9,17.9852814 11.0147186,20 13.5,20 Z M13.5,21 C10.4624339,21 8,18.5375661 8,15.5 C8,12.4624339 10.4624339,10 13.5,10 C16.5375661,10 19,12.4624339 19,15.5 C19,18.5375661 16.5375661,21 13.5,21 Z\\" fill-rule=\\"nonzero\\"></path><rect x=\\"13\\" y=\\"12\\" width=\\"1\\" height=\\"4\\"></rect><rect x=\\"13\\" y=\\"15\\" width=\\"4\\" height=\\"1\\"></rect></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">Item 2</div>
    </a>
</div><!--ITEM-->
</div><!--GROUP-->
</div><!--SECTION-->
    <div class=\\"hig__global-nav__sidenav__links\\">
        <!--LINK-->
    </div>
    <div class=\\"hig__global-nav__sidenav__copywright\\">© 2017 Autodesk, Inc. </div>
</div><!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__container__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;

exports[`<Group> children: <Item> can render a list of <Item> elements 1`] = `
"<div class=\\"hig__global-nav\\">
    <div class=\\"hig__global-nav__sidenav\\">
    <div class=\\"hig__global-nav__side-nav__section\\">
    <div class=\\"hig__global-nav__side-nav__section__header\\">
        <span class=\\"hig__global-nav__side-nav__section__header__label\\">Project</span>
        <span class=\\"hig__global-nav__side-nav__section__header__name\\">X</span>
    </div>
    <div class=\\"hig__global-nav__side-nav__section__group\\">
    <div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"21\\" viewBox=\\"0 0 20 21\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"project-management\\"><g transform=\\"translate(.242)\\" fill=\\"#34495E\\" fill-rule=\\"evenodd\\"><path d=\\"M15,8.15002717 C14.6738533,8.08382289 14.3400214,8.03878262 14,8.01640228 L14,1 L1,1 L1,17 L6.15002717,17 C6.21952655,17.3423796 6.31234911,17.6762897 6.42676427,18 L0,18 L0,0 L15,0 L15,8.15002717 Z M8.40093151,10 C8.07157864,10.3054873 7.76971094,10.6401947 7.49945072,11 L3,11 L3,10 L8.40093151,10 Z M6.42676427,13 C6.31234911,13.3237103 6.21952655,13.6576204 6.15002717,14 L3,14 L3,13 L6.42676427,13 Z M3,4 L12,4 L12,5 L3,5 L3,4 Z M3,7 L12,7 L12,8 L3,8 L3,7 Z\\" fill-rule=\\"nonzero\\"></path><path d=\\"M13.5,20 C15.9852814,20 18,17.9852814 18,15.5 C18,13.0147186 15.9852814,11 13.5,11 C11.0147186,11 9,13.0147186 9,15.5 C9,17.9852814 11.0147186,20 13.5,20 Z M13.5,21 C10.4624339,21 8,18.5375661 8,15.5 C8,12.4624339 10.4624339,10 13.5,10 C16.5375661,10 19,12.4624339 19,15.5 C19,18.5375661 16.5375661,21 13.5,21 Z\\" fill-rule=\\"nonzero\\"></path><rect x=\\"13\\" y=\\"12\\" width=\\"1\\" height=\\"4\\"></rect><rect x=\\"13\\" y=\\"15\\" width=\\"4\\" height=\\"1\\"></rect></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">Item 1</div>
    </a>
</div><div class=\\"hig__global-nav__side-nav__section__group__item\\">
    <a href=\\"#\\">
        <div class=\\"hig__global-nav__side-nav__section__group__item__icon\\"><div class=\\"hig__icon\\"><svg width=\\"20\\" height=\\"21\\" viewBox=\\"0 0 20 21\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"project-management\\"><g transform=\\"translate(.242)\\" fill=\\"#34495E\\" fill-rule=\\"evenodd\\"><path d=\\"M15,8.15002717 C14.6738533,8.08382289 14.3400214,8.03878262 14,8.01640228 L14,1 L1,1 L1,17 L6.15002717,17 C6.21952655,17.3423796 6.31234911,17.6762897 6.42676427,18 L0,18 L0,0 L15,0 L15,8.15002717 Z M8.40093151,10 C8.07157864,10.3054873 7.76971094,10.6401947 7.49945072,11 L3,11 L3,10 L8.40093151,10 Z M6.42676427,13 C6.31234911,13.3237103 6.21952655,13.6576204 6.15002717,14 L3,14 L3,13 L6.42676427,13 Z M3,4 L12,4 L12,5 L3,5 L3,4 Z M3,7 L12,7 L12,8 L3,8 L3,7 Z\\" fill-rule=\\"nonzero\\"></path><path d=\\"M13.5,20 C15.9852814,20 18,17.9852814 18,15.5 C18,13.0147186 15.9852814,11 13.5,11 C11.0147186,11 9,13.0147186 9,15.5 C9,17.9852814 11.0147186,20 13.5,20 Z M13.5,21 C10.4624339,21 8,18.5375661 8,15.5 C8,12.4624339 10.4624339,10 13.5,10 C16.5375661,10 19,12.4624339 19,15.5 C19,18.5375661 16.5375661,21 13.5,21 Z\\" fill-rule=\\"nonzero\\"></path><rect x=\\"13\\" y=\\"12\\" width=\\"1\\" height=\\"4\\"></rect><rect x=\\"13\\" y=\\"15\\" width=\\"4\\" height=\\"1\\"></rect></g></svg></div></div>
        <div class=\\"hig__global-nav__side-nav__section__group__item__title\\">Item 2</div>
    </a>
</div><!--ITEM-->
</div><!--GROUP-->
</div><!--SECTION-->
    <div class=\\"hig__global-nav__sidenav__links\\">
        <!--LINK-->
    </div>
    <div class=\\"hig__global-nav__sidenav__copywright\\">© 2017 Autodesk, Inc. </div>
</div><!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__container__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;
